### PR TITLE
[PE1-1292] manage OACs in distribution repository

### PR DESCRIPTION
**Pull request by LCaparelli**
## Description

Update the Distribution repository to delegate syncing and deleting OACs to the OAC repo.

## Motivation and Context

This is needed in order to support private S3 buckets as origins via OAC (Origin Access Control).

## How has this been tested?

Automated tests.

## Checklist

- [x] I have implemented automated tests for the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by LCaparelli** [comment]
> ## Description
> ## Motivation and Context
> ## How has this been tested?
> ## Checklist
> * [ ]  I have implemented automated tests for the changes.
> * [ ]  I have updated the documentation accordingly.

**Review by monicajss — APPROVED** [review]

File: internal/cloudfront/repository.go [hunk 1]
```diff
@@ -125,6 +139,10 @@ func (r repository) Create(d Distribution) (Distribution, error) {
 		return Distribution{}, fmt.Errorf("creating distribution: %v", err)
 	}
 
+	if err := r.createAOCs(d); err != nil {
```
Comment (go-faustino) [inline internal/cloudfront/repository.go:142]: nit: should be createOACs

File: internal/cloudfront/repository.go [hunk 2]
```diff
@@ -138,6 +156,19 @@ func (r repository) Sync(d Distribution) (Distribution, error) {
 		return Distribution{}, fmt.Errorf("getting distribution config: %v", err)
 	}
 
+	syncedOACs, err := r.syncAOCs(d, output)
```
Comment (go-faustino) [inline internal/cloudfront/repository.go:159]: nit: should be syncOACs

File: internal/cloudfront/repository.go [hunk 3]
```diff
@@ -192,7 +223,15 @@ func (r repository) Delete(d Distribution) error {
 		IfMatch: eTag,
 	}
 	_, err = r.cloudfrontCli.DeleteDistribution(input)
-	return cdnaws.IgnoreErrorCode(err, awscloudfront.ErrCodeNoSuchDistribution)
+	if cdnaws.IgnoreErrorCode(err, awscloudfront.ErrCodeNoSuchDistribution) != nil {
+		return err
+	}
+
+	if err := r.deleteAOCs(output.DistributionConfig); err != nil {
```
Comment (go-faustino) [inline internal/cloudfront/repository.go:230]: nit: should be deleteOACs

File: internal/cloudfront/repository.go [hunk 4]
```diff
@@ -192,7 +223,15 @@ func (r repository) Delete(d Distribution) error {
...
+		return fmt.Errorf("deleting AOCs: %v", err)
```
Comment (go-faustino) [inline internal/cloudfront/repository.go:231]: nit: should be deleting OACs

File: internal/cloudfront/repository.go [hunk 5]
```diff
@@ -265,3 +304,90 @@ func (r repository) distributionByID(id string) (*awscloudfront.GetDistributionO
 	}
 	return r.cloudfrontCli.GetDistribution(input)
 }
+func (r repository) createAOCs(d Distribution) error {
```
Comment (go-faustino) [inline internal/cloudfront/repository.go:307]: nit: should be createOACs

File: internal/cloudfront/repository.go [hunk 6]
```diff
@@ -265,3 +304,90 @@ func (r repository) distributionByID(id string) (*awscloudfront.GetDistributionO
...
+	for _, o := range d.OACs() {
+		if _, err := r.oacRepo.Sync(o); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r repository) syncAOCs(desired Distribution, observed *awscloudfront.GetDistributionConfigOutput) (synced []OAC, err error) {
```
Comment (go-faustino) [inline internal/cloudfront/repository.go:316]: nit: should be syncOACs

File: internal/cloudfront/repository.go [hunk 7]
```diff
@@ -265,3 +304,90 @@ func (r repository) distributionByID(id string) (*awscloudfront.GetDistributionO
...
+	toBeSynced, toBeDeleted := r.diffDesiredAndObservedAOCs(desired, observed.DistributionConfig)
```
Comment (go-faustino) [inline internal/cloudfront/repository.go:317]: nit: should be diffDesiredAndObservedOACs

File: internal/cloudfront/repository.go [hunk 8]
```diff
@@ -265,3 +304,90 @@ func (r repository) distributionByID(id string) (*awscloudfront.GetDistributionO
...
+		_, err := r.oacRepo.Delete(oac)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return toBeSynced, nil
+}
+
+func (r repository) deleteAOCs(cfg *awscloudfront.DistributionConfig) error {
```
Comment (go-faustino) [inline internal/cloudfront/repository.go:337]: nit: should be deleteOACs

File: internal/cloudfront/repository.go [hunk 9]
```diff
@@ -265,3 +304,90 @@ func (r repository) distributionByID(id string) (*awscloudfront.GetDistributionO
...
+	for _, o := range toBeDeleted {
+		if _, err := r.oacRepo.Delete(o); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r repository) diffDesiredAndObservedAOCs(desired Distribution, observed *awscloudfront.DistributionConfig) (toBeSynced []OAC, toBeDeleted []OAC) {
```
Comment (go-faustino) [inline internal/cloudfront/repository.go:351]: nit: should be diffDesiredAndObservedOACs

**Comment by LCaparelli** [comment]
> Approved, left a few nits on AOC vs OAC wink. Just so it doesn't confuse our future selves into thinking they're two different things.

Yup, me confusing this initially has had some long lasting effects lol thanks!

**Pull request merged**